android/wear-os-samples WearWidgets fails

wear-os-samples/WearWidget on  main [?] via 🅶 v9.4.1 via ☕ v26 via 🅺 on ☁️  yschimke@google.com took 2s 
❯ compose-preview render
compose-preview: plugin not applied in any build.gradle(.kts); running via auto-inject. For best IDE / agent support add to your module's plugins { } block: id("ee.schimke.composeai.preview") version "0.11.0" (suppress with --no-plugin-warning or COMPOSE_PREVIEW_NO_PLUGIN_WARNING=1).
FAILURE: Build failed with an exception.
* What went wrong:
> There were failing tests. See the report at: file:///Users/yschimke/workspace/wear-os-samples/WearWidget/app/build/reports/tests/composePreviewRender/index.html
* Try:
> Run with --scan to get full insights from a Build Scan (powered by Develocity).

Run with --verbose for full build output.

Failing tests in :app:composePreviewRender (4):
  ee.schimke.composeai.renderer.RobolectricRenderTest.renderPreview[RenderPreviewEntry(id=com.google.example.wear_widget.HelloWidgetKt.HelloWidgetPreview_PARAM_0, functionName=HelloWidgetPreview, className=com.google.example.wear_widget.HelloWidgetKt, sourceFile=src/main/java/com/google/example/wear_widget/HelloWidget.kt, params=RenderPreviewParams(name=null, device=null, widthDp=null, heightDp=null, density=2.625, fontScale=1.0, showSystemUi=false, showBackground=false, backgroundColor=0, uiMode=0, locale=null, group=null, wrapperClassName=null, previewParameterProviderClassName=com.google.example.wear_widget.WearWidgetParamsProviderSnapshot, previewParameterLimit=2147483647, kind=COMPOSE), captures=[RenderPreviewCapture(advanceTimeMillis=null, scroll=null, animation=null, focus=null, focusGif=null, ambient=null, renderOutput=renders/HelloWidgetKt.HelloWidgetPreview_PARAM_0.png, cost=1.0)], dataProducts=[])]
    Failed to interact with raw FileDescriptor internals; perhaps JRE has changed?
    java.lang.RuntimeException: Failed to interact with raw FileDescriptor internals; perhaps JRE has changed?
    at org.robolectric.interceptors.AndroidInterceptors$FileDescriptorInterceptor.setInt(AndroidInterceptors.java:88)
    at com.android.internal.os.ApplicationSharedMemory.create(ApplicationSharedMemory.java:94)
    at org.robolectric.android.internal.AndroidTestEnvironment.setUpApplicationState(AndroidTestEnvironment.java:229)
    at org.robolectric.RobolectricTestRunner.beforeTest(RobolectricTestRunner.java:307)
    at org.robolectric.internal.SandboxTestRunner.executeInSandbox(SandboxTestRunner.java:518)
  ee.schimke.composeai.renderer.RobolectricRenderTest.renderPreview[RenderPreviewEntry(id=com.google.example.wear_widget.HelloWidgetKt.HelloWidgetPreview_PARAM_1, functionName=HelloWidgetPreview, className=com.google.example.wear_widget.HelloWidgetKt, sourceFile=src/main/java/com/google/example/wear_widget/HelloWidget.kt, params=RenderPreviewParams(name=null, device=null, widthDp=null, heightDp=null, density=2.625, fontScale=1.0, showSystemUi=false, showBackground=false, backgroundColor=0, uiMode=0, locale=null, group=null, wrapperClassName=null, previewParameterProviderClassName=com.google.example.wear_widget.WearWidgetParamsProviderSnapshot, previewParameterLimit=2147483647, kind=COMPOSE), captures=[RenderPreviewCapture(advanceTimeMillis=null, scroll=null, animation=null, focus=null, focusGif=null, ambient=null, renderOutput=renders/HelloWidgetKt.HelloWidgetPreview_PARAM_1.png, cost=1.0)], dataProducts=[])]
    Failed to interact with raw FileDescriptor internals; perhaps JRE has changed?
    java.lang.RuntimeException: Failed to interact with raw FileDescriptor internals; perhaps JRE has changed?
    at org.robolectric.interceptors.AndroidInterceptors$FileDescriptorInterceptor.setInt(AndroidInterceptors.java:88)
    at com.android.internal.os.ApplicationSharedMemory.create(ApplicationSharedMemory.java:94)
    at org.robolectric.android.internal.AndroidTestEnvironment.setUpApplicationState(AndroidTestEnvironment.java:229)
    at org.robolectric.RobolectricTestRunner.beforeTest(RobolectricTestRunner.java:307)
    at org.robolectric.internal.SandboxTestRunner.executeInSandbox(SandboxTestRunner.java:518)
  ee.schimke.composeai.renderer.RobolectricRenderTest.renderPreview[RenderPreviewEntry(id=com.google.example.wear_widget.WeatherWidgetKt.WeatherWidgetPreview_PARAM_0, functionName=WeatherWidgetPreview, className=com.google.example.wear_widget.WeatherWidgetKt, sourceFile=src/main/java/com/google/example/wear_widget/WeatherWidget.kt, params=RenderPreviewParams(name=null, device=null, widthDp=null, heightDp=null, density=2.625, fontScale=1.0, showSystemUi=false, showBackground=false, backgroundColor=0, uiMode=0, locale=null, group=null, wrapperClassName=null, previewParameterProviderClassName=com.google.example.wear_widget.WearWidgetParamsProviderSnapshot, previewParameterLimit=2147483647, kind=COMPOSE), captures=[RenderPreviewCapture(advanceTimeMillis=null, scroll=null, animation=null, focus=null, focusGif=null, ambient=null, renderOutput=renders/WeatherWidgetKt.WeatherWidgetPreview_PARAM_0.png, cost=1.0)], dataProducts=[])]
    Failed to interact with raw FileDescriptor internals; perhaps JRE has changed?
    java.lang.RuntimeException: Failed to interact with raw FileDescriptor internals; perhaps JRE has changed?
    at org.robolectric.interceptors.AndroidInterceptors$FileDescriptorInterceptor.setInt(AndroidInterceptors.java:88)
    at com.android.internal.os.ApplicationSharedMemory.create(ApplicationSharedMemory.java:94)
    at org.robolectric.android.internal.AndroidTestEnvironment.setUpApplicationState(AndroidTestEnvironment.java:229)
    at org.robolectric.RobolectricTestRunner.beforeTest(RobolectricTestRunner.java:307)
    at org.robolectric.internal.SandboxTestRunner.executeInSandbox(SandboxTestRunner.java:518)
  ee.schimke.composeai.renderer.RobolectricRenderTest.renderPreview[RenderPreviewEntry(id=com.google.example.wear_widget.WeatherWidgetKt.WeatherWidgetPreview_PARAM_1, functionName=WeatherWidgetPreview, className=com.google.example.wear_widget.WeatherWidgetKt, sourceFile=src/main/java/com/google/example/wear_widget/WeatherWidget.kt, params=RenderPreviewParams(name=null, device=null, widthDp=null, heightDp=null, density=2.625, fontScale=1.0, showSystemUi=false, showBackground=false, backgroundColor=0, uiMode=0, locale=null, group=null, wrapperClassName=null, previewParameterProviderClassName=com.google.example.wear_widget.WearWidgetParamsProviderSnapshot, previewParameterLimit=2147483647, kind=COMPOSE), captures=[RenderPreviewCapture(advanceTimeMillis=null, scroll=null, animation=null, focus=null, focusGif=null, ambient=null, renderOutput=renders/WeatherWidgetKt.WeatherWidgetPreview_PARAM_1.png, cost=1.0)], dataProducts=[])]
    Failed to interact with raw FileDescriptor internals; perhaps JRE has changed?
    java.lang.RuntimeException: Failed to interact with raw FileDescriptor internals; perhaps JRE has changed?
    at org.robolectric.interceptors.AndroidInterceptors$FileDescriptorInterceptor.setInt(AndroidInterceptors.java:88)
    at com.android.internal.os.ApplicationSharedMemory.create(ApplicationSharedMemory.java:94)
    at org.robolectric.android.internal.AndroidTestEnvironment.setUpApplicationState(AndroidTestEnvironment.java:229)
    at org.robolectric.RobolectricTestRunner.beforeTest(RobolectricTestRunner.java:307)
    at org.robolectric.internal.SandboxTestRunner.executeInSandbox(SandboxTestRunner.java:518)

--- a/daemon/harness/src/main/kotlin/ee/schimke/composeai/daemon/harness/HarnessLauncher.kt
+++ b/daemon/harness/src/main/kotlin/ee/schimke/composeai/daemon/harness/HarnessLauncher.kt
@@ -295,6 +295,9 @@ class RealAndroidHarnessLauncher(
         add("--add-opens=java.base/java.lang=ALL-UNNAMED")
         add("--add-opens=java.base/java.lang.reflect=ALL-UNNAMED")
         add("--add-opens=java.base/java.nio=ALL-UNNAMED")
+        // SDK 36 + JDK 25/26: Robolectric's `FileDescriptorInterceptor.setInt`
+        // reflects through `jdk.internal.access.SharedSecrets` (see #1328).
+        add("--add-opens=java.base/jdk.internal.access=ALL-UNNAMED")
         // Render output directory.
         add("-Dcomposeai.render.outputDir=${rendersDir.absolutePath}")
         // Wire the manifest router (same as desktop).

--- a/gradle-plugin/src/main/kotlin/ee/schimke/composeai/plugin/AndroidPreviewClasspath.kt
+++ b/gradle-plugin/src/main/kotlin/ee/schimke/composeai/plugin/AndroidPreviewClasspath.kt
@@ -215,6 +215,19 @@ internal object AndroidPreviewClasspath {
       // opens. Reached via `PathIterator` — triggered here by Wear Compose's
       // curved text renderer.
       "--add-opens=java.base/java.nio=ALL-UNNAMED",
+      // Robolectric's `FileDescriptorInterceptor.setInt` reflects into
+      // `jdk.internal.access.SharedSecrets#getJavaIOFileDescriptorAccess()`
+      // to mutate `FileDescriptor.fd` (the older `Field.setInt` path was
+      // replaced in 4.13+). On SDK 36 sandboxes the framework's
+      // `com.android.internal.os.ApplicationSharedMemory.create()` runs
+      // during `AndroidTestEnvironment.setUpApplicationState`, hits the
+      // interceptor, and without this opens the JDK raises
+      // `IllegalAccessException: class … cannot access class
+      // jdk.internal.access.SharedSecrets (in module java.base) because
+      // module java.base does not export jdk.internal.access` — wrapped by
+      // Robolectric as `Failed to interact with raw FileDescriptor
+      // internals; perhaps JRE has changed?`. Issue #1328.
+      "--add-opens=java.base/jdk.internal.access=ALL-UNNAMED",
     )
 
   /**

--- a/gradle-plugin/src/main/kotlin/ee/schimke/composeai/plugin/AndroidPreviewSupport.kt
+++ b/gradle-plugin/src/main/kotlin/ee/schimke/composeai/plugin/AndroidPreviewSupport.kt
@@ -1439,12 +1439,7 @@ internal object AndroidPreviewSupport {
         useJUnit()
 
         jvmArgs(agpTestTask?.jvmArgs ?: emptyList<String>())
-        jvmArgs(
-          "--add-opens=java.base/java.io=ALL-UNNAMED",
-          "--add-opens=java.base/java.lang=ALL-UNNAMED",
-          "--add-opens=java.base/java.lang.reflect=ALL-UNNAMED",
-          "--add-opens=java.base/java.nio=ALL-UNNAMED",
-        )
+        jvmArgs(AndroidPreviewClasspath.buildJvmArgs())
         agpTestTask?.javaLauncher?.orNull?.let { javaLauncher.set(it) }
 
         systemProperty("robolectric.graphicsMode", "NATIVE")

--- a/gradle-plugin/src/test/kotlin/ee/schimke/composeai/plugin/AndroidPreviewClasspathTest.kt
+++ b/gradle-plugin/src/test/kotlin/ee/schimke/composeai/plugin/AndroidPreviewClasspathTest.kt
@@ -91,6 +91,18 @@ class AndroidPreviewClasspathTest {
     AndroidPreviewClasspath.validateApplicationOnClasspath(listOf(dir, notAJar, androidJar))
   }
 
+  @Test
+  fun `buildJvmArgs opens jdk_internal_access for FileDescriptorInterceptor`() {
+    // Regression for issue #1328: Robolectric 4.16's `FileDescriptorInterceptor.setInt`
+    // reflects into `jdk.internal.access.SharedSecrets`, which the JDK refuses to expose to
+    // an unnamed module without `--add-opens=java.base/jdk.internal.access=ALL-UNNAMED`. On
+    // SDK 36 sandboxes `ApplicationSharedMemory.create()` runs during Robolectric setup and
+    // hits that interceptor, surfacing as `Failed to interact with raw FileDescriptor
+    // internals; perhaps JRE has changed?`.
+    assertThat(AndroidPreviewClasspath.buildJvmArgs())
+      .contains("--add-opens=java.base/jdk.internal.access=ALL-UNNAMED")
+  }
+
   private fun writeAndroidJar(file: File) {
     writeJar(file, mapOf("android/app/Application.class" to ByteArray(16)))
   }

